streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasksTest.java

```diff
@@ -517,6 +519,147 @@ public void shouldCloseCleanlyWithSuspendedTaskAndEOS() {
         assignedTasks.shutdown(true);
     }
 
+    @Test
+    public void shouldClearZombieCreatedTasks() {
+        new TaskTestSuite() {
+            @Override
+            public void additionalSetup(final StreamTask task) {
+                task.close(false, true);
+            }
+
+            @Override
+            public void action(final StreamTask task) {
+                assignedTasks.addNewTask(task);
+            }
+
+            @Override
+            public Set<TaskId> taskIds() {
+                return assignedTasks.created.keySet();
+            }
+
+            @Override
+            public List<TopicPartition> expectedLostChangelogs() {
+                return clearingPartitions;
+            }
+        }.createTaskAndClear();
+    }
+
+    @Test
+    public void shouldClearZombieRestoringTasks() {
+        new TaskTestSuite() {
+            @Override
+            public void additionalSetup(final StreamTask task) {
+                EasyMock.expect(task.partitions()).andReturn(Collections.emptySet()).anyTimes();
+                task.closeStateManager(false);
```

Comment:
Hmm this is interesting: for created the running tasks we call its `task.close` function whereas for `restoring` we only call `task.closeStateManager`, is it intentional? If yes why? cc @ableegoldman

Reply:
I asked you exactly that a few months ago :) You referenced some old PR but basically the takeaway was, a restoring task hasn't initialized anything but its state, therefore needs to close only the state manager

Reply:
It's a little awkward, @abbccdda maybe we should at least put in a comment explaining this for now so we don't have to keep asking each other

Reply:
Ultimately I've been thinking all these different task types should literally be different types, so that we can just call `task.close` and have it close whatever is needed for a restoring task, or suspended, etc. We could probably do some further cleanup, for example on close all tasks call `suspend` then `closeSuspended` but only active & running tasks are ever just suspended -- and we could probably also drop the double checkpointing

Reply:
I think I agree, but my confusion is that why we call close on created tasks? Basically the weird thing I noticed is that for closing as zombies:

1. created : call `close`
2. restoring : call `closeStateMgr`
3. running : call `close`
4. suspended : call `closeSuspended`

3 and 4 makes sense to me, but 1/2 are a bit weird: for created tasks, we do not initialize topology nor state managers, but still we call `close`, whereas restoring ones (where we initialized state manager but not topology) we call `closeStateMgr`.

Did a closer look at the code base, I think the reason is that in `closeNonRunningTasks` (which is triggered in `onPartitionsRevoked`) we actually trigger the following:

```
task.close(false /* clean */, false /* isZombie */)
```

I.e. we treat it as an "unclean" close, and hence we won't write checkpoint, and most importantly, when `closeTopology` throws (which would be the case since it is not initialized at all) we would just ignore it. So "accidentally" this becomes correct.

This is not a great pattern, and I like @ableegoldman 's suggestion that we should consider making "close" call to be more robust in a follow-up PR, for now let's stay with what it is then.

Reply:
It's good to reveal these error-prone close call setups here. My take is that each function starting with `close` should be distinguished clearly in function comment what it tries to do. And we should avoid including > 1 boolean variables, as it's hard to reason about. What about cleanClose() and uncleanClose()? Also should we be more specific about word `zombie`? This seems to be a really vague term too.

Reply:
I think (and hope) the term `zombie` is not used in an unclear way -- but there might be some confusion in the translation from "clients-speak" to "streams-speak" since what the clients simply call "lost" we call "zombie task". Some other examples along that line are the confusion about what revoked means -- sometimes its the partitions given up temporarily, sometimes only the partitions that are being reassigned and permanently given up.

But clear code can always help with unclear terminology: as an example of some progress, we used to have both "suspended" and "zombie" standbys, neither of which **really** makes sense for a standby. We cleaned up that code in some recent PRs so that, I feel, is now quite clear

@abbccdda I also like the idea of `cleanClose`, `unCleanClose`, where each specific task can make it clear what it needs to do in each case. A nice side-effect there is we can clearly comment what clean (or unclean) means at the top of each method, rather than try to interpret every brief parameter description